src/Promitor.Agents.ResourceDiscovery/Graph/Authentication.cs

```diff
@@ -1,20 +1,55 @@
 ﻿using System.Threading.Tasks;
-using Microsoft.IdentityModel.Clients.ActiveDirectory;
+using Azure.Core;
+using Azure.Identity;
 using Microsoft.Rest;
+using Promitor.Agents.Core.Configuration.Authentication;
 
 namespace Promitor.Agents.ResourceDiscovery.Graph
 {
     public static class Authentication
     {
-        public static async Task<ServiceClientCredentials> GetServiceClientCredentialsAsync(string resource, string clientId, string clientSecret, string tenantId)
+        //public static async Task<ServiceClientCredentials> GetServiceClientCredentialsAsync(string resource, string clientId, string clientSecret, string tenantId)
+        //{
+        //    AuthenticationContext authContext = new AuthenticationContext($"https://login.microsoftonline.com/{tenantId}");
+
+        //    AuthenticationResult authResult = await authContext.AcquireTokenAsync(resource, new ClientCredential(clientId, clientSecret));
+
+        //    ServiceClientCredentials serviceClientCredentials = new TokenCredentials(authResult.AccessToken);
+
+        //    return serviceClientCredentials;
+        //}
+
+        /// <summary>
+        /// Gets a valid token using a Service Principal or a Managed Identity
+        /// </summary>
+        public static async Task<TokenCredentials> GetTokenCredentialsAsync(string resource, string tenantId, AuthenticationMode authenticationMode, string managedIdentityId = null, string clientId = null, string clientSecret = null)
         {
-            AuthenticationContext authContext = new AuthenticationContext($"https://login.microsoftonline.com/{tenantId}");
+            TokenCredential tokenCredential;
+
+            if (authenticationMode == AuthenticationMode.ServicePrincipal)
+            {
+                tokenCredential = new ClientSecretCredential(tenantId, clientId, clientSecret);
+            }
+            else if (!string.IsNullOrEmpty(managedIdentityId))
+            {
+                tokenCredential = new ManagedIdentityCredential(managedIdentityId);
+            }
+            else
+            {
+                tokenCredential = new DefaultAzureCredential();
+            }
+
+            if (!resource.ToLowerInvariant().EndsWith("/.default"))
```

Comment:
Can you elaborate why we need this?

Reply:
When you reaching an endpoint, using a impersonate identity, the only endpoint available is always '/.default' 
MSAL add the './default' string to your resource request behind the scene.
We have to do it here, since we are at a lower level (and we are not using MSAL; by the way)

Reply:
That's ok, thanks for explaining! Would you mind adding a comment with that please?